include/fence.h

```diff
@@ -0,0 +1,150 @@
+
+	friend bool operator==(const buffer_snapshot& lhs, const buffer_snapshot& rhs) { return lhs.m_sr == rhs.m_sr && lhs.m_data == rhs.m_data; }
+
+	friend bool operator!=(const buffer_snapshot& lhs, const buffer_snapshot& rhs) { return !operator==(lhs, rhs); }
+
+  private:
+	template <typename U, int Dims2>
+	friend class detail::buffer_fence_promise;
+
+	subrange<Dims> m_sr;
+	std::unique_ptr<T[]> m_data; // cannot use std::vector here because of vector<bool> m(
+
+	explicit buffer_snapshot(subrange<Dims> sr, std::unique_ptr<T[]> data) : m_sr(sr), m_data(std::move(data)) {}
+};
+
+} // namespace celerity::experimental
+
+namespace celerity::detail {
+
+template <typename T>
+class host_object_fence_promise : public detail::fence_promise {
+  public:
+	explicit host_object_fence_promise(const experimental::host_object<T>& obj) : m_host_object(obj) {}
+
+	std::future<T> get_future() { return m_promise.get_future(); }
+
+	void fulfill() override { m_promise.set_value(std::as_const(detail::get_host_object_instance(m_host_object))); }
+
+  private:
+	experimental::host_object<T> m_host_object;
+	std::promise<T> m_promise;
+};
+
+template <typename DataT, int Dims>
+class buffer_fence_promise : public detail::fence_promise {
+  public:
+	explicit buffer_fence_promise(const buffer<DataT, Dims>& buf, const subrange<Dims>& sr) : m_buffer(buf), m_subrange(sr) {}
+
+	std::future<experimental::buffer_snapshot<DataT, Dims>> get_future() { return m_promise.get_future(); }
+
+	void fulfill() override {
+		const auto access_info = runtime::get_instance().get_buffer_manager().get_host_buffer<DataT, Dims>(
+		    get_buffer_id(m_buffer), access_mode::read, range_cast<3>(m_subrange.range), id_cast<3>(m_subrange.offset));
+		assert((access_info.offset <= m_subrange.offset) == id_cast<Dims>(id<3>(true, true, true)));
+		auto data = std::make_unique<DataT[]>(m_subrange.range.size());
+		memcpy_strided(access_info.buffer.get_pointer(), data.get(), sizeof(DataT), access_info.buffer.get_range(), m_subrange.offset - access_info.offset,
+		    m_subrange.range, {}, m_subrange.range);
+		m_promise.set_value(experimental::buffer_snapshot<DataT, Dims>(m_subrange, std::move(data)));
+	}
+
+  private:
+	buffer<DataT, Dims> m_buffer;
+	subrange<Dims> m_subrange;
+	std::promise<experimental::buffer_snapshot<DataT, Dims>> m_promise;
+};
+
+} // namespace celerity::detail
+
+namespace celerity::experimental {
+
+/**
+ * Asynchronously captures the value of a host object by copy, introducing the same dependencies as a side-effect would.
+ *
+ * Waiting on the returned future in the application thread can stall scheduling of more work. To hide latency, either submit more command groups between
+ * fence and wait operations or ensure that other independent command groups are eligible to run while the fence is executed.
+ */
+template <typename T>
+std::future<T> fence(celerity::distr_queue&, const experimental::host_object<T>& obj) {
```

Comment:
Creating a fence for a `host_object<T&>` currently doesn't work, and it probably shouldn't. The produced error is a bit obscure though - maybe add a static assertion here to catch this early?

Reply:
I've added a static assertion.

 `<T&>` and `<void>` are a bit of a stain anyway. We should overhaul host objects at some point.